cranelift/codegen/src/isa/x86/abi.rs

```diff
@@ -859,11 +1002,58 @@ fn insert_common_epilogue(
     pos: &mut EncCursor,
     reg_type: ir::types::Type,
     csrs: &RegisterSet,
+    fpr_slot: Option<&StackSlot>,
     isa: &dyn TargetIsa,
     is_last: bool,
     mut cfa_state: Option<CFAState>,
 ) {
     let word_size = isa.pointer_bytes() as isize;
+
+    // Even though instructions to restore FPRs are inserted first, we have to append them after
+    // restored GPRs to satisfy parameter order in the return.
+    let mut restored_fpr_values = alloc::vec::Vec::new();
+
+    // Restore FPRs before we move RSP and invalidate stack slots.
+    if csrs.iter(FPR).len() != 0 {
+        let mut fpr_offset = 0;
+
+        // `stack_store` is not directly encodable in x86_64 at the moment, so we'll need a base
+        // address. We are well after postopt could run, so load the CSR region base once here,
+        // instead of hoping that the addr/store will be combined later.
+        let stack_addr = pos.ins().stack_addr(
+            types::I64,
+            *fpr_slot.expect("if FPRs are preserved, a stack slot is allocated for them"),
+            0,
+        );
+
+        // At this point we won't be using volatile registers for anything except for return
+        // registers. Arbitrarily pick RBP as it won't hold an interesting value, and we're about
+        // to restore it at the end of the function anyway.
+        pos.func.locations[stack_addr] = ir::ValueLoc::Reg(RU::rbp as u16);
```

Comment:
Original comment here:
> This is brittle in the face of frame pointer elision. A more applicable register would be `r11`, but there are no rex encodings of `fstDisp8` or `fldDisp8`, so using `r11` as a base violates constraints that the registers must have index <=7.

If memory serves, rex encodings for these have landed, so I'll adjust this to `r11` and see how it goes :crossed_fingers:

Reply:
A status update: I added `infer_rex` to [`load` and `store`](https://github.com/bytecodealliance/wasmtime/blob/master/cranelift/codegen/meta/src/isa/x86/encodings.rs#L1828-L1844) but haven't yet got around to the other memory operations. Let me know if those missing encodings are needed and I'll try to add them next.

Reply:
yes, `r11` works perfectly here now - thanks!